### PR TITLE
Add support for extensionsType api 

### PR DESCRIPTION
**Pull request by Arif-lakhani**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/k8s-extension/azext_k8s_extension/custom.py [hunk 1]
```diff
@@ -369,54 +369,159 @@ def delete_k8s_extension(
...
+    return client.location_list(
+        location,
+        plan_publisher,
+        plan_product,
+        plan_name,
+        release_train,
+        cluster_type)
+
+# get by location
+def get_extension_type_by_location(client, location, extension_type):
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/custom.py:393]: >get [](http://example.com/codeflow?start=4&length=3)

change this to 'show' to stay consistent with the terms used in CLI

File: src/k8s-extension/azext_k8s_extension/custom.py [hunk 2]
```diff
@@ -369,54 +369,159 @@ def delete_k8s_extension(
...
+
+    return client.location_list(
+        location,
+        plan_publisher,
+        plan_product,
+        plan_name,
+        release_train,
+        cluster_type)
+
+# get by location
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/custom.py:392]: >get [](http://example.com/codeflow?start=2&length=3)

Change this to 'show'

File: src/k8s-extension/azext_k8s_extension/custom.py [hunk 3]
```diff
@@ -369,54 +369,159 @@ def delete_k8s_extension(
...
+        location, 
+        extension_type,
+        release_train,
+        cluster_type,
+        major_version,
+        show_latest)
     return versions_list
 
+# get version by location
+def get_extension_type_version_by_location(
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/custom.py:423]: Change 'get' to 'show'

File: src/k8s-extension/azext_k8s_extension/custom.py [hunk 4]
```diff
@@ -369,54 +369,159 @@ def delete_k8s_extension(
...
+        cluster_rp, 
+        cluster_type, 
+        cluster_name, 
+        plan_publisher, 
+        plan_product, 
+        plan_name, 
+        release_train)
+
+# get by cluster
+def get_extension_type_by_cluster(
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/custom.py:461]: 'show' instead of 'get'

File: src/k8s-extension/azext_k8s_extension/custom.py [hunk 5]
```diff
@@ -369,54 +369,159 @@ def delete_k8s_extension(
...
+        show_latest
+       )
 
-def show_k8s_cluster_extension_type(client, resource_group_name, cluster_type, cluster_name, extension_type):
-    """Get an existing Extension Type.
-    """
-    # Determine ClusterRP
-    cluster_rp, parent_api_version = get_cluster_rp_api_version(cluster_type)
+# get version by cluster 
+def get_extension_type_version_by_cluster(
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/custom.py:505]: 'show' instead of 'get'

File: src/k8s-extension/azext_k8s_extension/_format.py [hunk 6]
```diff
@@ -34,23 +34,28 @@ def __get_table_row(result):
...
     name = result['name']
     defaultScope, allowMultipleInstances, defaultReleaseNamespace = '', '', ''
-    if result['supportedScopes']:
-        defaultScope = result['supportedScopes']['defaultScope']
-        if result['supportedScopes']['clusterScopeSettings']:
-            allowMultipleInstances = result['supportedScopes']['clusterScopeSettings']['allowMultipleInstances']
-            defaultReleaseNamespace = result['supportedScopes']['clusterScopeSettings']['defaultReleaseNamespace']
+    if result['properties']['supportedScopes']:
+        defaultScope = result['properties']['supportedScopes']['defaultScope']
+        if result['properties']['supportedScopes']['clusterScopeSettings'] is not None:
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/_format.py:54]: >if result['properties']['supportedScopes']['clusterScopeSettings'] is not None: [](http://example.com/codeflow?start=8&length=79)

Any reason why you are not handling 'namespacescopesettings'?

File: src/k8s-extension/azext_k8s_extension/_help.py [hunk 7]
```diff
@@ -83,41 +83,79 @@
     short-summary: Commands to discover Kubernetes Extension Types.
 """
 
-helps[f'{consts.EXTENSION_NAME} extension-types list'] = f"""
+helps[f'{consts.EXTENSION_NAME} extension-types get-by-cluster'] = f"""
     type: command
-    short-summary: List Kubernetes Extension Types.
+    short-summary: Get properties for a Cluster Extension Type for an existing cluster.
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/_help.py:88]: >for an existing cluster [](http://example.com/codeflow?start=63&length=23)

Expand this some more to clarify how the cluster information is used in filtering - I believe it is by the clusterType AND the region?  Are there other properties of the cluster used in filtering the results?

File: src/k8s-extension/azext_k8s_extension/_help.py [hunk 8]
```diff
@@ -83,41 +83,79 @@
...
+    short-summary: Get properties for a Cluster Extension Type in a region.
+    examples:
+      - name: Get properties for a Cluster Extension Type in a region
+        text: |-
+          az {consts.EXTENSION_NAME} extension-types get-by-location --location eastus --extension-type microsoft.openservicemesh
+"""
+
+helps[f'{consts.EXTENSION_NAME} extension-types get-version-by-cluster'] = f"""
+    type: command
+    short-summary: Get properties associated with a Cluster Extension Type version for an existing cluster.
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/_help.py:107]: >xisting cluster [](http://example.com/codeflow?start=91&length=15)

Here too, clarify which properties of the cluster are used in filtering the versions

File: src/k8s-extension/azext_k8s_extension/_help.py [hunk 9]
```diff
@@ -83,41 +83,79 @@
...
+    examples:
+      - name: Get properties associated with a Cluster Extension Type version in a region.
+        text: |-
+          az {consts.EXTENSION_NAME} extension-types list --location eastus --extension-type microsoft.openservicemesh \
+--version 1.0.0
+"""
+
+helps[f'{consts.EXTENSION_NAME} extension-types list-by-cluster'] = f"""
+    type: command
+    short-summary: List available Cluster Extension Types for an existing cluster.
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/_help.py:127]: >existing cluster [](http://example.com/codeflow?start=65&length=16)

Here too - clarify the cluster properties used for filtering

File: src/k8s-extension/azext_k8s_extension/_help.py [hunk 10]
```diff
@@ -83,41 +83,79 @@
...
         text: |-
-          az {consts.EXTENSION_NAME} extension-types list-by-location --location eastus2euap
+          az {consts.EXTENSION_NAME} extension-types list --location eastus
 """
 
-helps[f'{consts.EXTENSION_NAME} extension-types show'] = f"""
+helps[f'{consts.EXTENSION_NAME} extension-types list-versions-by-cluster'] = f"""
     type: command
-    short-summary: Show properties for a Kubernetes Extension Type.
+    short-summary: List available versions for a Cluster Extension Type for a given cluster.
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/_help.py:146]: >given cluster [](http://example.com/codeflow?start=78&length=13)

Same as above comments - clarify pl.

File: src/k8s-extension/azext_k8s_extension/_params.py [hunk 11]
```diff
@@ -99,32 +99,34 @@ def load_arguments(self, _):
...
+                   options_list=['--plan-product'],
+                   help='The plan product is referring to the Marketplace Product ID of the extension.')
+        c.argument('plan_publisher',
+                   arg_group="Marketplace",
+                   options_list=['--plan-publisher'],
+                   help='The plan publisher is referring to the Marketplace Publisher ID of the extension')        
+        c.argument('major_version',
+                   help='Filter results by the major version of an extension type. For example if 1 is specified, all versions with major version 1 (1.1, 1.1.2) will be shown. The default value is None')        
+        c.argument('show_latest',
+                   help='Filter results by only the latest version. For example, if this flag is used the latest version of the extensionType will be shown.')
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/_params.py:132]: >  [](http://example.com/codeflow?start=101&length=1)

Add 'only' here

File: src/k8s-extension/azext_k8s_extension/vendored_sdks/models.py [hunk 12]
```diff
@@ -6,4 +6,4 @@
 # --------------------------------------------------------------------------
 from .v2022_01_15_preview.models import *
 from .v2022_04_02_preview.models import *
-from .v2022_11_01.models import *
+from .v2023_05_01.models import *
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/vendored_sdks/models.py:9]: >.v2023_05_01. [](http://example.com/codeflow?start=5&length=13)

I am assuming you have not made any changes to the generated files under vendored_sdks folder, and so not reviewing these files.  If you have made any changes, let me know and I will review them

File: src/k8s-extension/azext_k8s_extension/_format.py (hunk 6 above)
Comment (Arif-lakhani) [reply src/k8s-extension/azext_k8s_extension/_format.py:54]: I just tweaked an existing function to work with the current data format. Not sure if we want namespace scoped settings to be viewed in the table format

File: src/k8s-extension/azext_k8s_extension/_help.py (hunk 7 above)
Comment (Arif-lakhani) [reply src/k8s-extension/azext_k8s_extension/_help.py]: Done

File: src/k8s-extension/azext_k8s_extension/_format.py (hunk 6 above)
Comment (NarayanThiru) [reply src/k8s-extension/azext_k8s_extension/_format.py:54]: It'd be good to be consistent.  Can you pl. add that too?

**Comment by NarayanThiru** [comment]
> 

It'd be good to move these above Marketplace group to keep them together

---
Refers to: src/k8s-extension/azext_k8s_extension/_params.py:133 in c2ac205. [](commit_id = c2ac2052c84229cbd87719e6198e892240c61783, deletion_comment = False)

**Review by NarayanThiru — APPROVED** [review]
:shipit:

**Pull request merged**